tachyon/zk/transcript/poseidon_transcript.h

```diff
@@ -0,0 +1,120 @@
+#ifndef TACHYON_ZK_TRANSCRIPT_POSEIDON_TRANSCRIPT_H_
+#define TACHYON_ZK_TRANSCRIPT_POSEIDON_TRANSCRIPT_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/buffer/buffer.h"
+#include "tachyon/base/buffer/vector_buffer.h"
+#include "tachyon/crypto/hashes/sponge/poseidon/halo2_poseidon.h"
+#include "tachyon/math/elliptic_curves/affine_point.h"
+#include "tachyon/zk/transcript/transcript.h"
+
+namespace tachyon::zk {
+
+template <typename AffinePointTy>
+class PoseidonRead : public TranscriptRead<AffinePointTy> {
+ public:
+  using ScalarField = typename AffinePointTy::ScalarField;
+  using Buffer = typename tachyon::base::Buffer;
+
+  constexpr PoseidonRead() = default;
+  explicit PoseidonRead(Buffer read_buf)
+      : state_(tachyon::crypto::PoseidonConfig<ScalarField>::CreateCustom(
+            size_t{8}, uint64_t{5}, size_t{8}, size_t{63}, size_t{0})),
+        reader_(std::move(read_buf)) {}
+
+  constexpr bool ReadPoint(AffinePointTy& point_out) override {
+    if (reader_.Read(&point_out)) {
+      return CommonPoint(point_out);
+    } else {
+      return false;
+    }
+  }
+
+  constexpr bool ReadScalar(ScalarField& scalar_out) override {
+    if (reader_.Read(&scalar_out)) {
+      return CommonScalar(scalar_out);
+    } else {
+      return false;
+    }
+  }
+
+  constexpr Challenge255<AffinePointTy> SqueezeChallenge() override {
+    return Challenge255<AffinePointTy>(state_.SqueezeNativeFieldElements(1)[0]);
+  }
+
+  constexpr bool CommonPoint(const AffinePointTy& point) override {
+    std::vector<ScalarField> coords = {
+        ScalarField::FromBigInt(point.x().ToBigInt()),
+        ScalarField::FromBigInt(point.y().ToBigInt())};
+    return state_.Absorb(coords);
+  }
+
+  constexpr bool CommonScalar(const ScalarField& scalar) override {
+    return state_.Absorb(scalar);
+  }
+
+ private:
+  tachyon::crypto::Halo2PoseidonSponge<ScalarField> state_;
+  Buffer reader_;
+};
+
+template <typename AffinePointTy>
+class PoseidonWrite : public TranscriptWrite<AffinePointTy> {
+ public:
+  using BaseField = typename AffinePointTy::BaseField;
+  using ScalarField = typename AffinePointTy::ScalarField;
+  using VectorBuffer = tachyon::base::VectorBuffer;
+
+  constexpr PoseidonWrite() = default;
+  // Initialize a transcript given an output buffer.
+  explicit PoseidonWrite(VectorBuffer write_buf)
```

Comment:
I thought it was a viable change, but I am getting such error on macos
```
./tachyon/zk/transcript/poseidon_transcript.h:81:49: error: constexpr constructor's 1st parameter type 'tachyon::zk::PoseidonWrite::VectorBuffer' (aka 'tachyon::base::VectorBuffer') is not a literal type
  constexpr explicit PoseidonWrite(VectorBuffer write_buf)
                                   ~~~~~~~~~~~~~^~~~~~~~~
./tachyon/base/buffer/vector_buffer.h:11:22: note: 'VectorBuffer' is not literal because it is not an aggregate and has no constexpr constructors other than copy or move constructors
class TACHYON_EXPORT VectorBuffer : public Buffer {
```
I'll not fix this as you suggested.